src/app/fdctl/run/tiles/fd_replay.c

```diff
@@ -0,0 +1,731 @@
+      stake_weights_msg[1] = stake_weight_idx; /* staked_cnt */
+      stake_weights_msg[2] = fd_epoch_slot0( &epoch_bank->epoch_schedule, stake_weights_msg[0] ); /* start_slot */
+      stake_weights_msg[3] = epoch_bank->epoch_schedule.slots_per_epoch; /* slot_cnt */
+      FD_LOG_WARNING(("Sending stake weights %lu %lu %lu %lu", stake_weights_msg[0], stake_weights_msg[1], stake_weights_msg[2], stake_weights_msg[3]));
+      ulong tspub = fd_frag_meta_ts_comp( fd_tickcount() );
+
+      ulong stake_weights_sz = 4*sizeof(ulong) + (stake_weight_idx * sizeof(fd_stake_weight_t));
+      ulong stake_weights_sig = 4UL;
+      fd_mcache_publish( ctx->stake_weights_out_mcache, ctx->stake_weights_out_depth, ctx->stake_weights_out_seq, stake_weights_sig, ctx->stake_weights_out_chunk,
+        stake_weights_sz, 0UL, tsorig, tspub );
+      ctx->stake_weights_out_seq   = fd_seq_inc( ctx->stake_weights_out_seq, 1UL );
+      ctx->stake_weights_out_chunk = fd_dcache_compact_next( ctx->stake_weights_out_chunk, stake_weights_sz, ctx->stake_weights_out_chunk0, ctx->stake_weights_out_wmark );
+    }
+  }
+}
+
+static void
+unprivileged_init( fd_topo_t *      topo,
+                   fd_topo_tile_t * tile,
+                   void *           scratch ) {
+  // if( FD_UNLIKELY( tile->in_cnt != 1 ||
+  //                  topo->links[ tile->in_link_id[ STORE_IN_IDX     ] ].kind != FD_TOPO_LINK_KIND_STORE_TO_REPLAY ) ) {
+  //   FD_LOG_ERR(( "replay tile has none or unexpected input links %lu %lu %lu",
+  //                tile->in_cnt, topo->links[ tile->in_link_id[ 0 ] ].kind, topo->links[ tile->in_link_id[ 1 ] ].kind ));
+  // }
+  
+  if( FD_UNLIKELY( tile->out_link_id_primary == ULONG_MAX ) )
+    FD_LOG_ERR(( "replay tile missing a primary output link" ));
+
+  fd_flamenco_boot( NULL, NULL );
+  /* Scratch mem setup */
+  FD_SCRATCH_ALLOC_INIT( l, scratch );
+  fd_replay_tile_ctx_t * ctx = FD_SCRATCH_ALLOC_APPEND( l, alignof(fd_replay_tile_ctx_t), sizeof(fd_replay_tile_ctx_t) );
+  ulong scratch_top = FD_SCRATCH_ALLOC_FINI( l, 1UL );
+  if( FD_UNLIKELY( scratch_top > (ulong)scratch + scratch_footprint( tile ) ) )
+    FD_LOG_ERR(( "scratch overflow %lu %lu %lu", scratch_top - (ulong)scratch - scratch_footprint( tile ), scratch_top, (ulong)scratch + scratch_footprint( tile ) ));
+  
+  ctx->wksp = topo->workspaces[ topo->objs[ tile->tile_obj_id ].wksp_id ].wksp;
+
+  void * alloc_shmem = fd_wksp_alloc_laddr( ctx->wksp, fd_alloc_align(), fd_alloc_footprint(), 3UL );
+  if( FD_UNLIKELY( !alloc_shmem ) ) { 
+    FD_LOG_ERR(( "fd_alloc too large for workspace" )); 
+  }
+
+  ulong blockstore_obj_id = fd_pod_queryf_ulong( topo->props, ULONG_MAX, "blockstore" );
+  FD_TEST( blockstore_obj_id!=ULONG_MAX );
+  ctx->blockstore_wksp = topo->workspaces[ topo->objs[ blockstore_obj_id ].wksp_id ].wksp;
+
+  if( ctx->blockstore_wksp == NULL ) {
+    FD_LOG_ERR(( "no blocktore workspace" ));
+  }
+
+  ctx->last_stake_weights_push_time = 0;
+
+  /* Set up store tile input */
```

Comment:
Let's do all the non-link init first, and this goes after to maintain consistency with repair

Reply:
Done